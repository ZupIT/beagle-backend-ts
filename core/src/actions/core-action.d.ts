import { ActionFunction } from '../model/action';
export declare const createCoreAction: <Props = any>(name: string) => ActionFunction<Props>;
