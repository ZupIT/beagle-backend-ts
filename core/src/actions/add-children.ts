import { Expression } from '..'
import { Component } from '../model/component'
import { createCoreAction } from './core-action'

interface AddChildrenParams {
  /**
   * The id of the component to receive the new elements.
   */
  componentId: string,
  /**
   * The array of components (nodes) to be added.
   */
  value: Expression<Component[]>,
  /**
   * The mode to attach the new nodes. The default value is APPEND.
   * - APPEND: adds the new nodes to the end of the list of children.
   * - PREPEND: adds the new nodes to the start of the list of children.
   * - REPLACE: replaces the entire array of children by the new nodes.
   */
  mode?: 'APPEND' | 'PREPEND' | 'REPLACE',
}

/**
 * Modifies the current component tree by adding a set of elements to an existing node. A simple example would be to
 * add a new row to the interface whenever a button is clicked:
 *
 * ```tsx
 * <>
 *   <Container id="inputGroup">
 *     <TextInput placeholder="name" />
 *   </Container>
 *   <Button onPress={addChildren({ componentId: 'inputGroup', value: [<TextInput placeholder="name" />] })}>
 *     Add more
 *   </Button>
 * </>
 * ```
 *
 * @param params the parameters for the action: componentId, value and mode. See {@link AddChildrenParams}.
 * @returns an instance of Action
 */
export const addChildren = createCoreAction<AddChildrenParams>('addChildren')
