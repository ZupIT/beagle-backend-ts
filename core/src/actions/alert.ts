import { isDynamicExpression } from '../utils'
import { Actions, Action, ActionProps } from '../model/action'
import { Expression } from '../types'
import { createCoreAction } from './core-action'

export interface AlertParams {
  /**
   * The title of the dialog. It doesn't work on web platforms.
   */
  title?: Expression<string>,
  /**
   * The text to show inside the dialog.
   */
  message: Expression<string>,
  /**
   * The label of the button to close the dialog. It doesn't work on web platforms.
   *
   * @defaultValue `'OK'`
   */
  labelOk?: string,
  /**
   * Actions to run when the OK button is pressed.
   */
  onPressOk?: Actions,
}

type AlertProps = ActionProps<AlertParams>

const alertAction = createCoreAction<AlertParams>('alert')

interface AlertFunction {
  /**
   * Opens up a system dialog box to show an alert message to the user.
   *
   * @param message the text to show inside the dialog.
   * @returns an instance of Action
   */
  (message: AlertProps['message']): Action,
  /**
   * Opens up a system dialog box to show an alert message to the user.
   *
   * @param options the options for the dialog box: title, message, labelOk and onPressOk. See {@link AlertParams}.
   * @returns an instance of Action
   */
  (options: AlertProps): Action,
}

/** @category Actions */
export const alert: AlertFunction = args => (typeof args === 'string' || isDynamicExpression(args))
  ? alertAction({ message: args as Expression<string> })
  : alertAction(args as ActionProps<AlertParams>)
