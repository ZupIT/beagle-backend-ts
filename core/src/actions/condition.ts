import { Actions } from '../model/action'
import { Expression } from '../types'
import { createCoreAction } from './core-action'

interface ConditionParams {
  condition: Expression<boolean>,
  onTrue?: Actions,
  onFalse?: Actions,
}

export const condition = createCoreAction<ConditionParams>('condition')
