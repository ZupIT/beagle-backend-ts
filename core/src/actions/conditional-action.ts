import { Actions } from '../model/action'
import { DynamicExpression } from '../types'
import { createCoreAction } from './core-action'

export interface ConditionalActionParams {
  /**
   * The condition to verify.
   */
  condition: DynamicExpression<boolean>,
  /**
   * The actions to run if the condition turns out to be true.
   */
  onTrue?: Actions,
  /**
   * The actions to run if the condition turns out to be false.
   */
  onFalse?: Actions,
}

/**
 * Runs a set of actions depending on the value of "condition", which is a value based on a Context or an Operation.
 *
 * @category Actions
 * @param params the action parameters: condition, onTrue and onFalse. See {@link ConditionalActionParams}.
 * @returns an instance of Action
 */
export const conditionalAction = createCoreAction<ConditionalActionParams>('condition')
