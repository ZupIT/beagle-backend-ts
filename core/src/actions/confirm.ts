import { ContextNode, Expression } from '..'
import { Actions, ActionInterface, ActionProps } from '../model/action'
import { createCoreAction } from './core-action'

interface Confirm {
  title?: Expression<string>,
  message: Expression<string>,
  onPressOk?: Actions,
  onPressCancel?: Actions,
  labelOk?: string,
  labelCancel?: string,
}

type ConfirmProps = ActionProps<Confirm>

const confirmAction = createCoreAction<Confirm>('confirm')

interface AlertFunction {
  (message: ConfirmProps['message']): ActionInterface,
  (options: ConfirmProps): ActionInterface,
}

export const confirm: AlertFunction = args => (typeof args === 'string' || args instanceof ContextNode)
  ? confirmAction({ message: args })
  : confirmAction(args as ActionProps<Confirm>)
