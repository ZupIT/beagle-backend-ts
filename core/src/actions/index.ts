export { addChildren } from './add-children'
export { alert } from './alert'
export { condition } from './condition'
export { confirm } from './confirm'
export * from './navigation'
export { sendRequest, SendRequestParams } from './send-request'
export { setContext } from './set-context'
export { submitForm } from './submit-form'
