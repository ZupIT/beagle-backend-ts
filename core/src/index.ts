export { Component, WithContext, WithChildren } from './model/component'
export { Expression, HttpMethod, DynamicExpression } from './types'
export { RootContext } from './model/context/root-context'
export { coreNamespace } from './constants'
export { Actions, WithAnalytics, createAction } from './model/action'
export { AnyContextNode } from './model/context/types'
export { ContextNode } from './model/context/context-node'
export { createContext } from './model/context'
export { serialize } from './serializer'
export { FC, ComponentProps } from './jsx/types'
export { logger } from './logger'
export { isDynamicExpression } from './utils'
export { BeagleJSX } from './jsx'

import { Component } from './model/component'


declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace JSX {
    type Element = Component

    interface ElementChildrenAttribute {
      children: 'children',
    }

    interface IntrinsicElements {
      component: Element,
    }
  }
}
