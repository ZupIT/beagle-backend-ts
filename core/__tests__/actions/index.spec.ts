import { mapValues } from 'lodash'
import * as actions from 'src/actions'

describe('Actions module', () => {
  it('should export', () => expect(mapValues(actions, v => typeof v)).toMatchSnapshot())
})
