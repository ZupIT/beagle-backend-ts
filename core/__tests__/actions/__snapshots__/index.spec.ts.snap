// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Actions module should export 1`] = `
Object {
  "addChildren": "function",
  "alert": "function",
  "conditionalAction": "function",
  "confirm": "function",
  "openExternalUrl": "function",
  "openNativeRoute": "function",
  "popStack": "function",
  "popToView": "function",
  "popView": "function",
  "pushStack": "function",
  "pushView": "function",
  "request": "function",
  "resetApplication": "function",
  "resetStack": "function",
  "sendRequest": "function",
}
`;
